exact.go

```diff
@@ -3,104 +3,224 @@
+func (f Frac) IsZero() bool {
+	return f.P.Cmp(zero) == 0
+}
+
+// add but ignores sign
+func add(a, b Frac) Frac {
+	p1 := big.NewInt(0).Mul(a.P, b.Q)
+	p2 := big.NewInt(0).Mul(b.P, a.Q)
+	return Frac{
+		P: big.NewInt(0).Add(p1, p2),
+		Q: big.NewInt(0).Mul(a.Q, b.Q),
+	}
+}
+
 // Add two rationale numbers a and b.
 func Add(a, b Frac) Frac {
+	if a.Sign == b.Sign {
+		// -a + (-b) == (-a)-b == -(a+b)
+		// +a + (+b) == +(a+b)
+		r := add(NewFrac2(a.P, a.Q, false), NewFrac2(b.P, b.Q, false))
+		r.Sign = a.Sign
+		return r
+	}
+
+	at := big.NewInt(0).Mul(a.P, b.Q)
+	bt := big.NewInt(0).Mul(b.P, a.Q)
+
+	p := big.NewInt(0)
+	var sign bool
+
+	m := max(at, bt)
+	if m == at {
+		p = p.Sub(at, bt)
+		sign = a.Sign
+	} else {
+		p = p.Sub(bt, at)
+		sign = b.Sign
+	}
 	return Frac{
-		P: a.P*b.Q + b.P*a.Q,
-		Q: a.Q * b.Q,
+		Sign: sign,
+		P:    p,
+		Q:    big.NewInt(0).Mul(a.Q, b.Q),
 	}
 }
 
+func max(a, b *big.Int) *big.Int {
+	if a.Cmp(b) > 0 {
+		return a
+	}
+	return b
+}
+
 // Sub subtract the rational numbers a and b.
 func Sub(a, b Frac) Frac {
-	return Add(a, Frac{
-		P: -b.P,
-		Q: b.Q,
-	})
+	// a - b = a + (-b)
+	b.Sign = !b.Sign
+	return Add(a, b)
 }
 
 // Mul multiplies a and b.
 func Mul(a, b Frac) Frac {
-	return Frac{
-		P: a.P * b.P,
-		Q: a.Q * b.Q,
+	r := Frac{
+		Sign: a.Sign && b.Sign,
+		P:    big.NewInt(0).Mul(a.P, b.P),
+		Q:    big.NewInt(0).Mul(a.Q, b.Q),
 	}
+
+	return r
 }
 
 // Div divides a and b
 func Div(a, b Frac) Frac {
+	if b.P.Cmp(zero) == 0 || b.Q.Cmp(zero) == 0 {
+		panic("division by zero")
+	}
+	// a = p/q
+	// b = p'/q'
+	// a/b = (p/q)/(p'/q') = (p/q)*(q'/p')
 	return Mul(a, Frac{
-		P: b.Q,
-		Q: b.P,
+		Sign: a.Sign && b.Sign,
+		P:    b.Q,
+		Q:    b.P,
 	})
 }
 
-// gcd is the greatest common dividor
-func gcd(a, b int64) int64 {
-	if b == 0 {
-		return a
+// Abs returns the absolute value of x
+func Abs(x Frac) Frac {
+	if x.Sign {
+		x.Sign = false
+		return x
 	}
-	return gcd(b, a%b)
+	return x
 }
 
-// Simplify the fraction
-func Simplify(a Frac) Frac {
-	cd := gcd(a.P, a.Q) // common divisor
-	return Frac{
-		P: a.P / cd,
-		Q: a.Q / cd,
+// Lt is the less than (<) comparator.
+func Lt(a, b Frac) bool {
+	a, b = a.Simplify(), b.Simplify()
+	if a.Sign == b.Sign &&
+		a.Q.Cmp(b.Q) == 0 {
+		return a.P.Cmp(b.P) < 0
 	}
+
+	r := Sub(a, b)
+	return r.Sign
 }
 
 // Cmp compares if a equals b.
 func Cmp(a, b Frac) bool {
-	as := Simplify(a)
-	bs := Simplify(b)
-	return as.P == bs.P &&
-		as.Q == bs.Q
-}
-
-func reverse(s []rune) {
-	if len(s) < 2 {
-		return
-	}
-	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
-		s[i], s[j] = s[j], s[i]
+	if a.P.Cmp(zero) == 0 &&
+		b.P.Cmp(zero) == 0 {
+		return true
 	}
+	as := a.Simplify()
+	bs := b.Simplify()
+	return as.P.Cmp(bs.P) == 0 &&
+		as.Q.Cmp(bs.Q) == 0
 }
 
-// integer to []rune
-func itors(i int64, runeBase rune) []rune {
-	if i == 0 {
-		return []rune{runeBase}
-	}
-
+// text to []rune
+func textToRunes(num string) []rune {
```

Comment:
Not sure if I missed something, but I think you can just do:

```go
return []rune(num)
```
Go will do the heavy lifting

Reply:
Did not see that at the time, but you're obviously correct. 
This was a dumb rewrite of the old `itors` that handled the conversion to a different unicode base., but now for default base, we don't need it.
Thanks!